### PR TITLE
Files added

### PR DESCRIPTION
**Pull request by Jithsaavvy**
Please review the code and let me know if any changes are to be made.

### CODE REVIEW

File: CSP_Minizinc_Code.py [hunk 1]
```diff
@@ -0,0 +1,179 @@
+#Adaptive Deployment of Safety Monitors 
+
+from abc import ABC,abstractmethod
+from custom_dtypes import SafetyMonitor,Sensors,Platforms
```
Comment (AshAswin) [inline CSP_Minizinc_Code.py]: Rather import *

File: CSP_Minizinc_Code.py [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+
+from abc import ABC,abstractmethod
+from custom_dtypes import SafetyMonitor,Sensors,Platforms
+from minizinc import Instance, Model, Solver
+
+#Provide the contextual information needed for selection and deployment.
+class Context_Monitor:
+    def __init__(self,repo_image):
+        #Initially the robot's context monitor is null
+        self.__gripper_status = None
```
Comment (AshAswin) [inline CSP_Minizinc_Code.py]: I suppose gripper status. robot in motion, finger open to be bool. Better to initialize them to false than none. Set the time stamp to 0.00

File: CSP_Minizinc_Code.py [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+        #Robot senses the current state of its environment
+        #Initially the sensed data is provided explicitly
+        self.__gripper_status=True
+        self.__robot_in_motion=False
+        self.__finger_open=True
+
+    def update_info_to_repo(self):
+        #Updates the current state of the robot to the repository
+        self.__repo_image.update_context(self.__gripper_status,self.__robot_in_motion,self.__finger_open)
+
```
Comment (AshAswin) [inline CSP_Minizinc_Code.py:27]: Is this not an indentation error?

File: CSP_Minizinc_Code.py [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+    @abstractmethod
+    def update_repository(self):
+        pass
+
+#Contains definition for selecting the safety monitor based on the current context. The selected safety monitor information is updated in the repository.
+class Safety_Monitor_Selector(Selector):
+    def __init__(self,repo_image):
+        #Initially the attributes are set to null and no safety monitor is selected by default.
+        self.__current_context = None
```
Comment (AshAswin) [inline CSP_Minizinc_Code.py]: This is a list of three elements, not one. We have to come up with a convention to intialize bools. I would prefer everything to be intialized to False. Give your opinion

File: custom_dtypes.py [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+    FORCE_TACTILE_SENSORS = 3
+    NO_SELECTION = 4
+
+#The need for this has to be verified
+class Platforms(Enum):
+    INTEL = 1
+    AMD = 2
+    NXP = 3
+    NVIDIA = 4
+    NO_SELECTION = 5
```
Comment (AshAswin) [inline custom_dtypes.py]: Please add an endline

**Review by AshAswin — CHANGES_REQUESTED** [review]
Please address the review comments. We may need to have another round of review before merging with master

**Comment by Jithsaavvy** [comment]
Please find the updated files based on reviews. Kindly review it and let me know for any changes.
[CSP_Minizinc_updated_files.zip](https://github.com/AshAswin/Adaptive-Deployment-of-Safety-Monitors-of-Autonomous-Systems/files/4774613/CSP_Minizinc_updated_files.zip)

File: CSP_Minizinc_Code.py [hunk 6]
```diff
@@ -105,26 +105,26 @@ def __init__(self, repo_image):
...
-        self.__platform = SafetyMonitor.NO_SELECTION
+        self.__platform = Platforms.NO_SELECTION
 
     def query_repository(self):
         self.__current_safety_monitor = self.__repo_image.get_active_safety_monitor()
 
     def select_deployment_platform(self):
         #Obtains output from minizinc regarding the selected platform 
-        selected_platform = self.platform_selected()
+        __selected_platform = self.__platform_selected()
```
Comment (AshAswin) [inline CSP_Minizinc_Code.py]: __ before selected_platform variable not requried

File: custom_dtypes.py [hunk 7]
```diff
@@ -19,4 +18,6 @@ class Platforms(Enum):
     AMD = 2
     NXP = 3
     NVIDIA = 4
-    NO_SELECTION = 5
\ No newline at end of file
+    NO_SELECTION = 5
+
+
```
Comment (AshAswin) [inline custom_dtypes.py]: Remove this extra line

**Review by AshAswin — CHANGES_REQUESTED** [review]
Please address the review comments. Also, confirm whether this code executes.

File: CSP_Minizinc_Code.py [hunk 8]
```diff
@@ -37,6 +38,12 @@ def __init__(self, active_safety_monitor = SafetyMonitor.NO_SELECTION, active_se
         self.__safety_monitor_platform = safety_monitor_platform
         self.__current_context = current_context
         self.__fixed_deployment = fixed_deployment
+        self.name = 1
+        self.min_force_sensor_count = 1
+        self.min_tactile_sensor_count = 100
+        self.min_memory_fused = 400
+        self.np = 4
```
Comment (AshAswin) [inline CSP_Minizinc_Code.py]: Shall we better name this as n_platforms? If yes, then the same change has to be made in all usages

File: CSP_Minizinc_Code.py [hunk 9]
```diff
@@ -37,6 +38,12 @@ def __init__(self, active_safety_monitor = SafetyMonitor.NO_SELECTION, active_se
         self.__safety_monitor_platform = safety_monitor_platform
         self.__current_context = current_context
         self.__fixed_deployment = fixed_deployment
+        self.name = 1
```
Comment (AshAswin) [inline CSP_Minizinc_Code.py]: id would be more appropriate

File: CSP_Minizinc_Code.py [hunk 10]
```diff
@@ -37,6 +38,12 @@ def __init__(self, active_safety_monitor = SafetyMonitor.NO_SELECTION, active_se
         self.__safety_monitor_platform = safety_monitor_platform
         self.__current_context = current_context
         self.__fixed_deployment = fixed_deployment
+        self.name = 1
+        self.min_force_sensor_count = 1
+        self.min_tactile_sensor_count = 100
+        self.min_memory_fused = 400
+        self.np = 4
+        self.platforms = set([1,2,3,4])
```
Comment (AshAswin) [inline CSP_Minizinc_Code.py]: Why set is required here?

File: CSP_Minizinc_Code.py [hunk 11]
```diff
@@ -3,44 +3,48 @@
...
-                 current_context = [False, False, False],fixed_deployment = True,
+                 current_context = [False, False],fixed_deployment = True,
                  safety_monitor_platform = Platforms.NO_SELECTION):
 
         self.__active_safety_monitor = active_safety_monitor
         self.__active_sensor = active_sensor
         self.__safety_monitor_platform = safety_monitor_platform
         self.__current_context = current_context
         self.__fixed_deployment = fixed_deployment
+        self.min_force_sensor_count = 1
```
Comment (AshAswin) [inline CSP_Minizinc_Code.py]: This is actually presence of force sensor but we changed it to min_force_sensor_count to make our MiniZn model working. We'll have to look for possibilities to modify this

File: CSP_Minizinc_Code.py [hunk 12]
```diff
@@ -3,44 +3,48 @@
...
+        self.min_tactile_sensor_count = 100
+        self.min_memory_fused = 400
+        self.np = 4
+        self.platforms = set([1,2,3,4])
 
     #Setter methods for updating current context, safety monitor and platform information for deployment from other componenents
-    def update_context(self,gripper_status:bool,robot_motion:bool,finger_open=bool):
-        self.__current_context = gripper_status,robot_motion,finger_open
+    def update_context(self,gripper_status:bool,robot_motion:bool):
```
Comment (AshAswin) [inline CSP_Minizinc_Code.py]: We have to stick to a single convention with respect to specifying the argument datatypes (for example, gripper_status:bool). I find argument datatypes not to be mentioned in most of the code. Please remove argument datatype here or add in other places

File: CSP_Minizinc_Code.py [hunk 13]
```diff
@@ -75,18 +79,18 @@ def update_repository(self):
 class Safety_Monitor_Selector(Selector):
     def __init__(self,repo_image):
         #Initially the attributes are set to null and no safety monitor is selected by default.
-        self.__current_context = (False,False,False)
+        self.__current_context = (False,False)
```
Comment (AshAswin) [inline CSP_Minizinc_Code.py]: Its better to add a comment what current context contains (.i.e. Robot in motion and gripper status)

File: CSP_Minizinc_Code.py [hunk 14]
```diff
@@ -95,49 +99,54 @@ def select_safety_monitor(self):
     def update_repository(self):
         self.__repo_image.update_current_safety_monitor(self.__selected_safety_monitor)
 
+    #Method for returning the selected safety monitor for performing unit test
```
Comment (AshAswin) [inline CSP_Minizinc_Code.py]: Unit Test methods have to be removed as per our last discussion with Prof

File: CSP_Minizinc_Code.py [hunk 15]
```diff
@@ -95,49 +99,54 @@ def select_safety_monitor(self):
     def update_repository(self):
         self.__repo_image.update_current_safety_monitor(self.__selected_safety_monitor)
 
+    #Method for returning the selected safety monitor for performing unit test
+    def get_safety_monitor_unittest(self):
+        return self.__selected_safety_monitor
+
 
 #Contains definition for selecting the platform based on the selected safety monitor. The selected platform information is updated in the repository.
```
Comment (AshAswin) [inline CSP_Minizinc_Code.py]: Is it possible to use a better commenting style? Please refer some code bases. TensorFlow for example

File: CSP_Minizinc_Code.py [hunk 16]
```diff
@@ -95,49 +99,54 @@ def select_safety_monitor(self):
...
 #The requirements are formulated into suitable constraints and platforms that satisfy the given constraints will be selected.
 class Platform_Selector(Selector):
-    def __init__(self, repo_image):
+    def __init__(self, repo_image, minizinc_model):
         #Initially the attributes are set to null and no safety monitor and platform is selected by default.
         self.__current_safety_monitor = SafetyMonitor.NO_SELECTION
         self.__repo_image = repo_image
         self.__platform = Platforms.NO_SELECTION
+        self.__minizinc_model = minizinc_model
+        self._selected_platform = None
```
Comment (AshAswin) [inline CSP_Minizinc_Code.py]: Double underscore missing

File: CSP_Minizinc_Code.py [hunk 17]
```diff
@@ -153,6 +162,10 @@ def __platform_selected(self):
 
 if __name__ == '__main__':
     
+    args = argparse.ArgumentParser("Desciption: Please include the constraint satisfaction minizn solver file ")
```
Comment (AshAswin) [inline CSP_Minizinc_Code.py]: Spell check

File: Test_CSP_Minizinc_Code.py [hunk 18]
```diff
@@ -0,0 +1,24 @@
...
+    context_monitor=Context_Monitor(repo)
+    context_monitor.get_robot_status()
+    context_monitor.update_info_to_repo()
+    safety_monitor=Safety_Monitor_Selector(repo)
+    safety_monitor.query_repository()
+    safety_monitor.select_safety_monitor()
+    current_safety_monitor=str(safety_monitor.get_safety_monitor_unittest())
+ 
+
+    def test_select_safety_monitor(self):
```
Comment (AshAswin) [inline Test_CSP_Minizinc_Code.py]: The Unit Test is not required for the same reason mentioned above

**Review by AshAswin — CHANGES_REQUESTED** [review]
Please address the review comments.

File: CSP_Minizinc_Code.py (hunk 10 above)
Comment (Manojkl) [reply CSP_Minizinc_Code.py]: Because that's the standard way of passing a argument into a minizinc environment.

File: CSP_Minizinc_Code.py (hunk 9 above)
Comment (Manojkl) [reply CSP_Minizinc_Code.py]: It's a outdated.

File: Test_CSP_Minizinc_Code.py (hunk 18 above)
Comment (Manojkl) [reply Test_CSP_Minizinc_Code.py]: We won't be using unittest anymore.

File: CSP_Minizinc_Code.py [hunk 19]
```diff
@@ -5,28 +5,57 @@
...
-#Provide the contextual information needed for selection and deployment.
+
 class Context_Monitor:
+    """
+    Provide the contextual information needed for selection and deployment.
+    """
     def __init__(self,repo_image):
-        #Initially the robot's context monitor is null
+        """
+        Initially the robot's context monitor is null
```
Comment (AshAswin) [inline CSP_Minizinc_Code.py]: This comment is not required

**Pull request merged**